File: docx_parser_converter_typescript/src/docx_parsers/styles/run_properties_parser.ts
import {
  extractElement,
  extractAttribute,
  extractBooleanAttribute,
  safeInt,
  // WORDML_NAMESPACE_PREFIX, // Not directly used as prefixes are part of keys
  DEFAULT_ATTRIBUTES_GROUP_NAME,
} from '../../helpers/common_helpers';
import {
  convertHalfPointsToPoints,
  convertTwipsToPoints,
} from '../../utils';
import {
  RunStylePropertiesModel,
  FontPropertiesModel,
  LanguagePropertiesModel,
  // Ensure other necessary models are imported if RunStylePropertiesModel is extended
} from '../../models/styles_models';

/**
 * Parses the <w:rPr> (Run Properties) element from DOCX XML.
 * @param rPrElement The <w:rPr> XML element object, or undefined.
 * @param attributesGroupName The key used by fast-xml-parser for the attributes group.
 * @returns A RunStylePropertiesModel object or undefined if no properties are found or input is invalid.
 */
export function parseRunProperties(
  rPrElement: any | undefined,
  attributesGroupName: string = DEFAULT_ATTRIBUTES_GROUP_NAME
): RunStylePropertiesModel | undefined {
  if (!rPrElement) {
    return undefined;
  }

  const props: Partial<RunStylePropertiesModel> = {};

  // Fonts (w:rFonts)
  const rFontsElement = extractElement(rPrElement, 'w:rFonts');
  if (rFontsElement) {
    const fontProps: Partial<FontPropertiesModel> = {};
    const ascii = extractAttribute(rFontsElement, 'w:ascii', attributesGroupName);
    const hAnsi = extractAttribute(rFontsElement, 'w:hAnsi', attributesGroupName);
    const eastAsia = extractAttribute(rFontsElement, 'w:eastAsia', attributesGroupName);
    const cs = extractAttribute(rFontsElement, 'w:cs', attributesGroupName);
    if (ascii) fontProps.ascii = ascii;
    if (hAnsi) fontProps.hAnsi = hAnsi;
    if (eastAsia) fontProps.eastAsia = eastAsia;
    if (cs) fontProps.cs = cs;
    if (Object.keys(fontProps).length > 0) {
      props.fonts = FontPropertiesModel.parse(fontProps);
    }
  }

  // Font Size (w:sz) - in half-points
  const szElement = extractElement(rPrElement, 'w:sz');
  if (szElement) {
    const val = extractAttribute(szElement, 'w:val', attributesGroupName);
    const sizeHalfPts = safeInt(val);
    if (sizeHalfPts !== undefined) {
      props.size_pt = convertHalfPointsToPoints(sizeHalfPts);
    }
  }
  // Consider w:szCs (complex script font size) if necessary

  // Color (w:color)
  const colorElement = extractElement(rPrElement, 'w:color');
  if (colorElement) {
    const val = extractAttribute(colorElement, 'w:val', attributesGroupName);
    if (val) props.color = val;
  }

  // Bold (w:b)
  const bElement = extractElement(rPrElement, 'w:b');
  if (bElement !== undefined) { // Check for presence of the element itself
    props.bold = extractBooleanAttribute(bElement, 'w:val', attributesGroupName);
  }


  // Italic (w:i)
  const iElement = extractElement(rPrElement, 'w:i');
  if (iElement !== undefined) {
    props.italic = extractBooleanAttribute(iElement, 'w:val', attributesGroupName);
  }

  // Underline (w:u)
  const uElement = extractElement(rPrElement, 'w:u');
  if (uElement) {
    const val = extractAttribute(uElement, 'w:val', attributesGroupName);
    if (val) props.underline = val;
  }

  // Strikethrough (w:strike and w:dstrike)
  const strikeElement = extractElement(rPrElement, 'w:strike');
  if (strikeElement !== undefined) {
    props.strike = extractBooleanAttribute(strikeElement, 'w:val', attributesGroupName);
  }
  const dstrikeElement = extractElement(rPrElement, 'w:dstrike');
  if (dstrikeElement !== undefined) {
     // If dstrike is present and true, it implies strike is also true
    if (extractBooleanAttribute(dstrikeElement, 'w:val', attributesGroupName)) {
        props.strike = true;
    }
  }

  // Hidden (w:vanish) - maps to RunStylePropertiesModel.hidden if such a field exists
  const vanishElement = extractElement(rPrElement, 'w:vanish');
  if (vanishElement !== undefined) {
    props.hidden = extractBooleanAttribute(vanishElement, 'w:val', attributesGroupName);
  }

  // Language (w:lang)
  const langElement = extractElement(rPrElement, 'w:lang');
  if (langElement) {
    const langProps: Partial<LanguagePropertiesModel> = {};
    const val = extractAttribute(langElement, 'w:val', attributesGroupName);
    const eastAsiaLang = extractAttribute(langElement, 'w:eastAsia', attributesGroupName);
    const bidiLang = extractAttribute(langElement, 'w:bidi', attributesGroupName);
    if (val) langProps.val = val;
    if (eastAsiaLang) langProps.eastAsia = eastAsiaLang;
    if (bidiLang) langProps.bidi = bidiLang;
    if (Object.keys(langProps).length > 0) {
      props.language = LanguagePropertiesModel.parse(langProps);
    }
  }

  // Highlight (w:highlight)
  const highlightElement = extractElement(rPrElement, 'w:highlight');
  if (highlightElement) {
    const val = extractAttribute(highlightElement, 'w:val', attributesGroupName);
    if (val) props.highlight = val;
  }

  // Shading (w:shd) - maps to RunStylePropertiesModel.shading_fill if such a field exists
  const shdElement = extractElement(rPrElement, 'w:shd');
  if (shdElement) {
    const fill = extractAttribute(shdElement, 'w:fill', attributesGroupName);
    const color = extractAttribute(shdElement, 'w:color', attributesGroupName);
    if (fill) props.shading_fill = fill;
    if (color) props.shading_color = color; // Assuming shading_color field
  }

  // Vertical Alignment / Text Position (w:vertAlign or w:position)
  const vertAlignElement = extractElement(rPrElement, 'w:vertAlign');
  if (vertAlignElement) {
      const val = extractAttribute(vertAlignElement, 'w:val', attributesGroupName);
      if (val) props.vert_align = val; // e.g., "superscript", "subscript", "baseline"
  }

  const positionElement = extractElement(rPrElement, 'w:position');
  if (positionElement) {
    const val = extractAttribute(positionElement, 'w:val', attributesGroupName);
    const posHalfPts = safeInt(val);
    if (posHalfPts !== undefined) {
      props.position_pt = convertHalfPointsToPoints(posHalfPts);
    }
  }

  // Kerning (w:kern) - in half-points
  const kernElement = extractElement(rPrElement, 'w:kern');
  if (kernElement) {
    const val = extractAttribute(kernElement, 'w:val', attributesGroupName);
    const kernHalfPts = safeInt(val);
    if (kernHalfPts !== undefined) {
      props.kerning_pt = convertHalfPointsToPoints(kernHalfPts);
    }
  }

  // Character Spacing (w:spacing) - in twips
  const spacingElement = extractElement(rPrElement, 'w:spacing');
  if (spacingElement) {
    const val = extractAttribute(spacingElement, 'w:val', attributesGroupName);
    const spacingTwips = safeInt(val);
    if (spacingTwips !== undefined) {
      props.character_spacing_twips = spacingTwips;
      // If model expects points: (props as any).character_spacing_pt = convertTwipsToPoints(spacingTwips);
    }
  }

  // Emboss (w:emboss)
  const embossElement = extractElement(rPrElement, 'w:emboss');
  if (embossElement !== undefined) {
    props.emboss = extractBooleanAttribute(embossElement, 'w:val', attributesGroupName);
  }

  // Outline (w:outline)
  const outlineElement = extractElement(rPrElement, 'w:outline');
  if (outlineElement !== undefined) {
    props.outline = extractBooleanAttribute(outlineElement, 'w:val', attributesGroupName);
  }

  // Shadow (w:shadow)
  const shadowElement = extractElement(rPrElement, 'w:shadow');
  if (shadowElement !== undefined) {
    props.shadow = extractBooleanAttribute(shadowElement, 'w:val', attributesGroupName);
  }

  // AllCaps (w:caps)
  const capsElement = extractElement(rPrElement, 'w:caps');
  if (capsElement !== undefined) {
    props.all_caps = extractBooleanAttribute(capsElement, 'w:val', attributesGroupName);
  }

  // SmallCaps (w:smallCaps)
  const smallCapsElement = extractElement(rPrElement, 'w:smallCaps');
  if (smallCapsElement !== undefined) {
    props.small_caps = extractBooleanAttribute(smallCapsElement, 'w:val', attributesGroupName);
  }

  if (Object.keys(props).length > 0) {
    try {
      return RunStylePropertiesModel.parse(props);
    } catch (error) {
      console.error("Error parsing RunStylePropertiesModel:", error, "Input props:", props, "Original rPrElement:", rPrElement);
      return undefined;
    }
  }
  return undefined;
}

File: docx_parser_converter_python/parsers/document/run_properties_parser.py
"""Parser for run properties elements."""

from lxml.etree import _Element as Element

from models.document.run import Language, RunFonts, RunProperties, Underline
from parsers.common.border_parser import parse_border
from parsers.common.color_parser import parse_color
from parsers.common.shading_parser import parse_shading
from parsers.utils import (
    find_child,
    get_attribute,
    get_int_attribute,
    parse_toggle,
)


def parse_run_fonts(element: Element | None) -> RunFonts | None:
    """Parse <w:rFonts> element.

    XML Example:
        <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial"/>

    Args:
        element: The <w:rFonts> element or None

    Returns:
        RunFonts model or None if element is None
    """
    if element is None:
        return None

    return RunFonts(
        ascii=get_attribute(element, "ascii"),
        h_ansi=get_attribute(element, "hAnsi"),
        east_asia=get_attribute(element, "eastAsia"),
        cs=get_attribute(element, "cs"),
        hint=get_attribute(element, "hint"),
    )


def parse_language(element: Element | None) -> Language | None:
    """Parse <w:lang> element.

    XML Example:
        <w:lang w:val="en-US" w:eastAsia="zh-CN" w:bidi="ar-SA"/>

    Args:
        element: The <w:lang> element or None

    Returns:
        Language model or None if element is None
    """
    if element is None:
        return None

    return Language(
        val=get_attribute(element, "val"),
        east_asia=get_attribute(element, "eastAsia"),
        bidi=get_attribute(element, "bidi"),
    )


def parse_underline(element: Element | None) -> Underline | None:
    """Parse <w:u> element.

    XML Example:
        <w:u w:val="single" w:color="FF0000"/>

    Args:
        element: The <w:u> element or None

    Returns:
        Underline model or None if element is None
    """
    if element is None:
        return None

    return Underline(
        val=get_attribute(element, "val"),
        color=get_attribute(element, "color"),
        theme_color=get_attribute(element, "themeColor"),
    )


def parse_run_properties(element: Element | None) -> RunProperties | None:
    """Parse <w:rPr> element.

    XML Example:
        <w:rPr>
            <w:rStyle w:val="Strong"/>
            <w:b/>
            <w:sz w:val="24"/>
            <w:color w:val="FF0000"/>
        </w:rPr>

    Args:
        element: The <w:rPr> element or None

    Returns:
        RunProperties model or None if element is None
    """
    if element is None:
        return None

    # Style reference
    r_style_elem = find_child(element, "rStyle")
    r_style = get_attribute(r_style_elem, "val") if r_style_elem is not None else None

    # Font settings
    r_fonts = parse_run_fonts(find_child(element, "rFonts"))

    # Boolean toggles
    b = parse_toggle(find_child(element, "b"))
    b_cs = parse_toggle(find_child(element, "bCs"))
    i = parse_toggle(find_child(element, "i"))
    i_cs = parse_toggle(find_child(element, "iCs"))
    caps = parse_toggle(find_child(element, "caps"))
    small_caps = parse_toggle(find_child(element, "smallCaps"))
    strike = parse_toggle(find_child(element, "strike"))
    dstrike = parse_toggle(find_child(element, "dstrike"))
    outline = parse_toggle(find_child(element, "outline"))
    shadow = parse_toggle(find_child(element, "shadow"))
    emboss = parse_toggle(find_child(element, "emboss"))
    imprint = parse_toggle(find_child(element, "imprint"))
    vanish = parse_toggle(find_child(element, "vanish"))
    spec_vanish = parse_toggle(find_child(element, "specVanish"))

    # Color
    color = parse_color(find_child(element, "color"))

    # Spacing/sizing
    spacing_elem = find_child(element, "spacing")
    spacing = get_int_attribute(spacing_elem, "val") if spacing_elem is not None else None

    w_elem = find_child(element, "w")
    w = get_int_attribute(w_elem, "val") if w_elem is not None else None

    kern_elem = find_child(element, "kern")
    kern = get_int_attribute(kern_elem, "val") if kern_elem is not None else None

    position_elem = find_child(element, "position")
    position = get_int_attribute(position_elem, "val") if position_elem is not None else None

    sz_elem = find_child(element, "sz")
    sz = get_int_attribute(sz_elem, "val") if sz_elem is not None else None

    sz_cs_elem = find_child(element, "szCs")
    sz_cs = get_int_attribute(sz_cs_elem, "val") if sz_cs_elem is not None else None

    # Highlight
    highlight_elem = find_child(element, "highlight")
    highlight = get_attribute(highlight_elem, "val") if highlight_elem is not None else None

    # Underline
    u = parse_underline(find_child(element, "u"))

    # Effect (deprecated)
    effect_elem = find_child(element, "effect")
    effect = get_attribute(effect_elem, "val") if effect_elem is not None else None

    # Border and shading
    bdr = parse_border(find_child(element, "bdr"))
    shd = parse_shading(find_child(element, "shd"))

    # Vertical alignment
    vert_align_elem = find_child(element, "vertAlign")
    vert_align = get_attribute(vert_align_elem, "val") if vert_align_elem is not None else None

    # Language
    lang = parse_language(find_child(element, "lang"))

    return RunProperties(
        r_style=r_style,
        r_fonts=r_fonts,
        b=b,
        b_cs=b_cs,
        i=i,
        i_cs=i_cs,
        caps=caps,
        small_caps=small_caps,
        strike=strike,
        dstrike=dstrike,
        outline=outline,
        shadow=shadow,
        emboss=emboss,
        imprint=imprint,
        vanish=vanish,
        color=color,
        spacing=spacing,
        w=w,
        kern=kern,
        position=position,
        sz=sz,
        sz_cs=sz_cs,
        highlight=highlight,
        u=u,
        effect=effect,
        bdr=bdr,
        shd=shd,
        vert_align=vert_align,
        lang=lang,
        spec_vanish=spec_vanish,
    )
